src/stack.cc

```diff
@@ -0,0 +1,395 @@
+				return cdada_stack_push_u<uint64_t>(m,
+								m->stack.u64,
+								val);
+			case 16:
+				return cdada_stack_push_u<cdada_u128_t>(m,
+								m->stack.u128,
+								val);
+			case 32:
+				return cdada_stack_push_u<cdada_u256_t>(m,
+								m->stack.u256,
+								val);
+			case 64:
+				return cdada_stack_push_u<cdada_u512_t>(m,
+								m->stack.u512,
+								val);
+			case 128:
+				return cdada_stack_push_u<cdada_u1024_t>(m,
+								m->stack.u1024,
+								val);
+			case 256:
+				return cdada_stack_push_u<cdada_u2048_t>(m,
+								m->stack.u2048,
+								val);
+			case 0:
+				CDADA_ASSERT(m->ops);
+				return (*m->ops->push)(m, val);
+			default:
+				break;
+		}
+	}catch(bad_alloc& e){
+		return CDADA_E_MEM;
+	}catch(...){}
+
+	CDADA_ASSERT(0);
+	return CDADA_E_UNKNOWN;
+}
+
+int cdada_stack_pop(cdada_stack_t* stack){
+
+	__cdada_stack_int_t* m = (__cdada_stack_int_t*)stack;
+
+	CDADA_CHECK_MAGIC(m);
+
+	try{
+		int c = m->ops? 0 : m->val_len;
+		switch(c){
+			case 1:
+				return cdada_stack_pop_u<uint8_t>(m,
+								m->stack.u8);
+			case 2:
+				return cdada_stack_pop_u<uint16_t>(m,
+								m->stack.u16);
+			case 4:
+				return cdada_stack_pop_u<uint32_t>(m,
+								m->stack.u32);
+			case 8:
+				return cdada_stack_pop_u<uint64_t>(m,
+								m->stack.u64);
+			case 16:
+				return cdada_stack_pop_u<cdada_u128_t>(m,
+								m->stack.u128);
+			case 32:
+				return cdada_stack_pop_u<cdada_u256_t>(m,
+								m->stack.u256);
+			case 64:
+				return cdada_stack_pop_u<cdada_u512_t>(m,
+								m->stack.u512);
+			case 128:
+				return cdada_stack_pop_u<cdada_u1024_t>(m,
+								m->stack.u1024);
+			case 256:
+				return cdada_stack_pop_u<cdada_u2048_t>(m,
+								m->stack.u2048);
+			case 0:
+				CDADA_ASSERT(m->ops);
+				return (*m->ops->pop)(m);
+			default:
+				break;
+		}
+	}catch(...){}
+
+	CDADA_ASSERT(0);
+	return CDADA_E_UNKNOWN;
+}
+
+uint32_t cdada_stack_size(const cdada_stack_t* stack){
+
+	__cdada_stack_int_t* m = (__cdada_stack_int_t*)stack;
+
+	if(unlikely(!m || m->magic_num != CDADA_MAGIC))
+		return 0;
+
+	try{
+		int c = m->ops? 0 : m->val_len;
+		switch(c){
+			case 1:
+				return m->stack.u8->size();
+			case 2:
+				return m->stack.u16->size();
+			case 4:
+				return m->stack.u32->size();
+			case 8:
+				return m->stack.u64->size();
+			case 16:
+				return m->stack.u128->size();
+			case 32:
+				return m->stack.u256->size();
+			case 64:
+				return m->stack.u512->size();
+			case 128:
+				return m->stack.u1024->size();
+			case 256:
+				return m->stack.u2048->size();
+			case 0:
+				CDADA_ASSERT(m->ops);
+				return (*m->ops->size)(m);
+			default:
+				break;
+		}
+	}catch(...){}
+
+	CDADA_ASSERT(0);
+	return 0;
+}
+
+int cdada_stack_top(const cdada_stack_t* stack, void* val){
+
+	__cdada_stack_int_t* m = (__cdada_stack_int_t*)stack;
+
+	CDADA_CHECK_MAGIC(m);
+
+	if(unlikely(!val))
+		return CDADA_E_INVALID;
+
+	//NOTE: we don't want std::stack insert "replace semantics", so we return
```

Comment:
Is this comment applicable to`stack->top()`

Reply:
No, I will remove it.

Reply:
Done.